keygen/src/ssl.c

```diff
@@ -0,0 +1,327 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <openssl/pem.h>
+#include <openssl/ec.h>
+#include <openssl/x509.h>
+#include <sodium.h>
+#include <errno.h>
+
+#include "ssl.h"
+#include "enftun/config.h"
+
+/**
+ * public_key_encode_x962() - Encode a public key from an EVP_PKEY structure
+ *
+ * Caller must free the returned buffer
+ *
+ * Return: A new buffer containing the x9.62-encoded string
+ */
+static char *public_key_encode_x962(EVP_PKEY *pkey)
+{
+    EC_KEY *key = NULL;
+    const EC_POINT * point = NULL;
+    char *hex = NULL;
+
+    key = EVP_PKEY_get1_EC_KEY(pkey);
+    if (!key) {
+        printf("%s Could not get EC_KEY from pkey", __func__);
+        goto out;
+    }
+
+    point = EC_KEY_get0_public_key(key);
+    if (!point) {
+        printf("%s Could not get point from EC_KEY", __func__);
+        goto out;
+    }
+
+    hex = EC_POINT_point2hex(EC_KEY_get0_group(key), point, EC_KEY_get_conv_form(key), NULL);
+
+out:
+    EC_KEY_free(key);
+    return hex;
+}
+
+/**
+ * gen_EC_PKEY() - Generate an Elliptic Curve key (NID_X9_62_prime256v1)
+ *
+ * Return: A pointer to a new EC_KEY instance or NULL on failure
+ */
+int gen_key(struct key *key_out)
+{
+    int ret;
+    EC_KEY * ec_key = NULL;
+
+    ec_key = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
+    if (!ec_key) {
+        fprintf(stderr, "%s ERR: Could not generate EC_KEY", __func__);
+        ret = 0;
+        goto out;
+    }
+
+    key_out->pkey = EVP_PKEY_new();
+    if (!key_out->pkey) {
+        fprintf(stderr, "%s ERR: Could not generate PKEY", __func__);
+        ret = 0;
+        goto cleanup;
+    }
+
+    ret = EC_KEY_generate_key(ec_key);
+    if (!ret) {
+        fprintf(stderr, "%s ERR: Could not generate EC_KEY", __func__);
+        ret = 0;
+        goto cleanup;
+    }
+
+    if (!EVP_PKEY_assign_EC_KEY(key_out->pkey, ec_key)) {
+        fprintf(stderr, "%s ERR: Could not assign EC_KEY to PKEY", __func__);
+        ret = 0;
+        goto cleanup;
+    }
+
+    key_out->hex_str = public_key_encode_x962(key_out->pkey);
+    if (!key_out->hex_str) {
+        fprintf(stderr, "%s ERR: Could not generate x862 public key from PKEY", __func__);
+        ret = 0;
+        /* Because EVP_PKEY_assign_EC_KEY succeeded both items will be freed when EC_KEY_free is run */
+        goto cleanup;
+    }
+
+    /* Success */
+    ret = 1;
+    goto out;
+
+cleanup:
+    if (ec_key)
+        EC_KEY_free(ec_key);
+    destroy_key(key_out);
+
+out:
+    return ret;
+}
+
+/**
+ * destroy_key() - Deallocate a key
+ *
+ * Return: None
+ */
+void destroy_key(struct key *key_in)
+{
+    if(key_in->pkey)
+        EVP_PKEY_free(key_in->pkey);
+    OPENSSL_free(key_in->hex_str);
```

Comment:
Need to check `hex_str` for `NULL`, too, right?

Reply:
`OPENSSL_free` is like `free` with NULL input. The documentation for `EVP_PKEY_free` specifies the same so I've removed that NULL check.

Reply:
I always forget that `free`'ing a `NULL` pointer is fine!